### PR TITLE
[MU4] Implemented dropdowns on notation page

### PR DESCRIPTION
**Pull request by Eism**
- Implemented StyledMenu component based on our popup
- Fixed toggling note input and added a toggling note input method
- Implemented dropdowns for add tuplets and add button on note input bar
- Implemented dropdown for toggling page view

https://user-images.githubusercontent.com/10116828/111128426-8bee7980-857d-11eb-9b52-d4dc808b2a74.mp4

### CODE REVIEW

**Comment by shoogle** [comment]
Is this to close #7645? If so, please add "Closes #7645" somewhere in the commit message (not necessarily on the top line).

Don't use "Fixes #7645" because that will get confusing with "Fix #7645" on the musescore.org tracker.

**Comment by Eism** [comment]
Closes #7670 
Closes #7650 
Closes #7676 
Closes #7651

**Comment by Eism** [comment]
@shoogle thanks, fixed

**Comment by Eism** [comment]
@shoogle oh, I deceived you
does this really need to be added in the commit message?
what is it for?

**Comment by shoogle** [comment]
> does this really need to be added in the commit message?

@Eism, maybe not. I was hoping "Closes #xxxx" in the commit message would do this automatically but now I'm not sure. Perhaps it links an issue but doesn't assign it to a project?

Anyway, if the commit message doesn't work (or is too much effort) then you need to click "Linked Issues" in the top right and link it to any issue that it fixes. Also click "Projects" and assign it to any projects that those issues are assigned to.



image

File: src/framework/uicomponents/qml/MuseScore/UiComponents/StyledMenuItem.qml [hunk 1]
```diff
@@ -1,89 +1,153 @@
...
+    enabled: Boolean(modelData) && Boolean(modelData.enabled)
 
-    property var hintIcon: checkable && checked ? IconCode.TICK : IconCode.NONE
-    property string shortcut: ""
+    isSelected: privateProperties.showedSubMenu != undefined || (privateProperties.hasIcon && privateProperties.isSelectable && privateProperties.isSelected)
 
-    background: Rectangle {
-        id: background
+    property var modelData
+    property bool reserveSpaceForInvisibleItems: true
```
Comment (RomanPudashkin) [inline src/framework/uicomponents/qml/MuseScore/UiComponents/StyledMenuItem.qml:17]: Can you explain please what is this needed for?

File: src/framework/uicomponents/qml/MuseScore/UiComponents/StyledMenuItem.qml [hunk 2]
```diff
@@ -1,89 +1,153 @@
...
     id: root
 
-    implicitHeight: 30
-    implicitWidth: 220
+    defaultColor: ui.theme.accentColor
+    enabled: Boolean(modelData) && Boolean(modelData.enabled)
 
-    property var hintIcon: checkable && checked ? IconCode.TICK : IconCode.NONE
-    property string shortcut: ""
+    isSelected: privateProperties.showedSubMenu != undefined || (privateProperties.hasIcon && privateProperties.isSelectable && privateProperties.isSelected)
```
Comment (RomanPudashkin) [inline src/framework/uicomponents/qml/MuseScore/UiComponents/StyledMenuItem.qml:14]: Better to use Boolean(privateProperties.showedSubMenu) here

File: src/framework/uicomponents/qml/MuseScore/UiComponents/StyledMenuItem.qml [hunk 3]
```diff
@@ -1,89 +1,153 @@
...
+import MuseScore.UiComponents 1.0
 import MuseScore.Ui 1.0
 
-MenuItem {
+ListItemBlank {
     id: root
 
-    implicitHeight: 30
-    implicitWidth: 220
+    defaultColor: ui.theme.accentColor
```
Comment (RomanPudashkin) [inline src/framework/uicomponents/qml/MuseScore/UiComponents/StyledMenuItem.qml:11]: Maybe is it better to rename this property to backgroundColor?

File: src/notation/view/noteinputbarmodel.cpp [hunk 4]
```diff
@@ -558,10 +598,227 @@ MenuItem NoteInputBarModel::makeActionItem(const ActionItem& action, const std::
...
+        return false;
+    }
+
+    return true;
+}
+
+void NoteInputBarModel::notifyAboutTupletItemChanged()
+{
+    int tupletItemIndex = itemIndex(TUPLET_ACTION_CODE);
+    if (tupletItemIndex == -1) {
```
Comment (RomanPudashkin) [inline src/notation/view/noteinputbarmodel.cpp:805]: It is better to add a static constant for this number (INVALID_INDEX), you use it in many places

File: src/notation/view/viewmodecontrolmodel.cpp [hunk 5]
```diff
@@ -22,77 +22,78 @@
...
-    : QAbstractListModel(parent)
+    : QObject(parent)
 {
-    m_viewModeOptions << ViewModeOption { qtrc("notation", "Page View"), "view-mode-page", ViewMode::PAGE }
-                      << ViewModeOption { qtrc("notation", "Continuous View"), "view-mode-continuous", ViewMode::LINE }
-                      << ViewModeOption { qtrc("notation", "Single Page"), "view-mode-single", ViewMode::SYSTEM };
 }
 
-void ViewModeControlModel::load()
+QVariant ViewModeControlModel::currentViewMode()
```
Comment (RomanPudashkin) [inline src/notation/view/viewmodecontrolmodel.cpp:39]: Please, add const to this method

File: src/notation/qml/MuseScore/NotationScene/NoteInputBar.qml [hunk 6]
```diff
@@ -46,7 +46,31 @@ Rectangle {
             height: gridView.cellWidth
 
             onClicked: {
-                noteInputModel.handleAction(item.codeRole)
+                if (menu.isOpened || item.showSubitemsByClickRole) {
+                    menu.toggleOpened()
+                    return
+                }
+
+                noteInputModel.handleAction(item.codeRole, -1)
```
Comment (RomanPudashkin) [inline src/notation/qml/MuseScore/NotationScene/NoteInputBar.qml:54]: Probably you needed a default value for the second argument in this method

**Review by RomanPudashkin — APPROVED** [review]

File: src/framework/uicomponents/qml/MuseScore/UiComponents/StyledMenuItem.qml (hunk 1 above)
Comment (Eism) [reply src/framework/uicomponents/qml/MuseScore/UiComponents/StyledMenuItem.qml:17]: By default, I reserve space for invisible items like shortcuts, icons, etc.
But this is not necessary for some popups, for example, page view control.
image

**Pull request merged**